### ISSUE TITLE
Hierarchical ls in `HSGP` / `HSGPPeriodic`

### ISSUE THREAD

**Issue by wd60622** [pymc-labs/pymc-marketing#1280]
No description provided.

**Comment by wd60622** [comment]
Have you done this logic @bwengals 
I tried and see that that the shapes out of `prior_linearized` need a bit of handling

**Comment by juanitorduz** [comment]
This would be great! I think we would need to make sure the spectral density expressions vecorize over `ls`? 

Actually, it seems the spectral density functions do handle vector length scales:

https://github.com/pymc-devs/pymc/blob/5d51953df0168fce4029675575f7002db71e4347/pymc/gp/cov.py#L574-L587

**Comment by wd60622** [comment]
> This would be great! I think we would need to make sure the spectral density expressions vecorize over `ls`? 
> 
> 
> 
> Actually, it seems the spectral density functions do handle vector length scales:
> 
> 
> 
> https://github.com/pymc-devs/pymc/blob/5d51953df0168fce4029675575f7002db71e4347/pymc/gp/cov.py#L574-L587

Yes, think we just need the dims to broadcast correctly. But i tried out and there needs a little thought behind the following operations with psd, etc

**Comment by bwengals** [comment]
@juanitorduz they do handle vector lengthscales, but in terms of input dimension. For example, a spatial GP with a different lengthscale for the x and the y directions.  It'd be really nice if they could handle both input dimension and the case below (vectorized, multi-dimensional inputs), but I had no idea how to do that.

There [are examples here](https://www.pymc.io/projects/examples/en/latest/gaussian_processes/HSGP-Advanced.html) of using a vector of lengthscales for the case where you have multiple GPs each with a different lengthscale that you're implementing in a vectorized way.   I'm purposefully saying they're multiple GP's because if the lengthscales are different then the kernels are different, and that's what differentiates one GP from another.  I'd bet for most hierarchical GPs they'd share the same lengthscale parameter. 

Which of these two cases are you trying to implement?

**Comment by juanitorduz** [comment]
Thanks @bwengals for the clarification. I think for MMMs you are right, all lengths scales would be similar. I think the main ingredients are in the HSGP advanced example you shared (which should be a similar strategy as in  https://www.pymc-labs.com/blog-posts/modelling-changes-marketing-effectiveness-over-time/ I believe)
